File: haskell/p-63.hs
{-# LANGUAGE UnicodeSyntax #-}

import Prelude.Unicode

data Tree a = Empty | Branch a (Tree a) (Tree a)
            deriving (Show, Eq)


completeBinaryTree ∷ Int → Tree Char
completeBinaryTree 0 = Empty
completeBinaryTree 1 = Branch 'x' Empty Empty
completeBinaryTree n =
    let sub = completeBinaryTree $ n - 1
    in Branch 'x' sub sub

File: prolog/p-63.pl
complete_binary_tree(0, nil) :- !.
complete_binary_tree(1, t(x, nil, nil)) :- !.
complete_binary_tree(N, Tree) :-
	N1 is N - 1,
	complete_binary_tree(N1, Sub),
	Tree = t(x, Sub, Sub).
